sdk/client/src/verifying_client.rs

```diff
@@ -0,0 +1,516 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    client::Client,
+    request::MethodRequest,
+    response::{MethodResponse, Response},
+    state::State,
+};
+use anyhow::{bail, ensure, format_err, Result};
+use diem_json_rpc_types::views::AccountView;
+use diem_types::{
+    account_address::AccountAddress, account_config::diem_root_address,
+    account_state::AccountState, account_state_blob::AccountStateWithProof,
+    epoch_change::EpochChangeProof, ledger_info::LedgerInfoWithSignatures,
+    on_chain_config::RegisteredCurrencies, proof::AccumulatorConsistencyProof,
+    transaction::Version, trusted_state::TrustedState, waypoint::Waypoint,
+};
+use std::{
+    collections::HashMap,
+    convert::{TryFrom, TryInto},
+    fmt::Debug,
+    sync::{Arc, RwLock},
+};
+
+// TODO(philiphayes): figure out retry strategy
+// TODO(philiphayes): real on-disk waypoint persistence
+// TODO(philiphayes): fill out rest of the methods
+// TODO(philiphayes): use real error types
+// TODO(philiphayes): non-verifying vs verifying client equivalence testing
+// TODO(philiphayes): all clients should validate chain id (allow users to trust-on-first-use or pre-configure)
+// TODO(philiphayes): we could abstract the async client so VerifyingClient takes a dyn Trait?
+
+// TODO(philiphayes): we should really add a real StateProof type (not alias) that
+// collects these types together. StateProof isn't a very descriptive name though...
+
+type StateProof = (
+    LedgerInfoWithSignatures,
+    EpochChangeProof,
+    AccumulatorConsistencyProof,
+);
```

Comment:
Yeah this might make more sense to have an actual type and have it maybe in the diem-types crate?